internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,156 @@
+	}
+	return nil
 }
 
-func (b *InMemoryWriteBuffer) WriteAt(data []byte, offset int64) error {
+// Allocates memory to flushedBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureFlushedBuffer() error {
+	if b.bufferSize == 0 {
+		return fmt.Errorf(ZeroSizeBufferError)
+	}
+	if b.flushed.buffer == nil {
+		b.flushed.buffer = make([]byte, 0, b.bufferSize)
+	}
 	return nil
 }
+
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, offset int64) error {
+	dataSize := int64(len(content))
+	receivedContentEndOffset := offset + dataSize
+	if dataSize == 0 {
+		return nil
+	}
+
+	// Ensure b.currentBuffer != nil
+	if err := b.ensureCurrentBuffer(); err != nil {
+		return err
+	}
+
+	// If receivedContentStartOffset is not in the range of offsets stored in current
+	// buffer, trigger temp file flow for writes.
+	if offset < b.current.offset.start ||
+		offset > b.current.offset.end {
+		// TODO: finalise write and trigger temp file flow.
+		logger.Debugf("TODO: finalise write and trigger temp file flow.")
+		return fmt.Errorf(NonSequentialWriteError)
+	}
+
+	// If data can be completely written to the current buffer, write it.
+	if receivedContentEndOffset <= b.current.offset.end {
+		return b.copyDataToBuffer(offset, receivedContentEndOffset, content)
+	}
+	// Else, write the chunk of data that can be written to the current buffer block.
+	l := b.current.offset.end - offset
+	err := b.copyDataToBuffer(offset, b.current.offset.end, content[:l])
+	if err != nil {
+		return err
+	}
+	//TODO: get status of last write to GCS. If unsuccessful, trigger temp file flow.
+	logger.Debugf("TODO: get status of last write to GCS and if " +
+		"unsuccessful, trigger temp file flow.")
+	if err = b.advanceToNextChunk(); err != nil {
+		return err
+	}
+	// TODO: trigger async upload to GCS for flushedBuffer.
+	logger.Debugf("TODO: trigger async upload to GCS for flushedBuffer.")
+
+	// Write the remaining data to currentBuffer.
+	return b.copyDataToBuffer(b.current.offset.start, receivedContentEndOffset, content[l:])
+}
+
+// Copies received content to currentChunk if it has capacity.
+func (b *InMemoryWriteBuffer) copyDataToBuffer(contentStartOffset, contentEndOffset int64, content []byte) error {
+	dataSize := int64(len(content))
+	si := contentStartOffset % b.bufferSize
+	ei := si + dataSize
+	if ei > b.bufferSize {
+		return fmt.Errorf(NotEnoughSpaceInCurrentBuffer)
+	}
+
+	copy(b.current.buffer[si:ei], content)
+	b.updateFileSize(contentEndOffset)
+	return nil
+}
+
+func (b *InMemoryWriteBuffer) advanceToNextChunk() error {
+	// ensure b.flushed.buffer != nil
+	if err := b.ensureFlushedBuffer(); err != nil {
+		return err
+	}
+
+	// Move current buffer to flushed buffer.
+	b.flushed.buffer, b.current.buffer = b.current.buffer, b.flushed.buffer
+	b.flushed.offset.start = b.current.offset.start
+	b.flushed.offset.end = b.current.offset.end
+
+	// Make current buffer ready for new writes coming from kernel.
+	clear(b.current.buffer)
```

Comment:
Why is this clear required ?

Reply:
current buffer will have previously written data. In case of random writes on current buffer, the old data will still exist if we don't clear.

Reply:
Right, but even if the current buffer will have previously written data, does it matter ? Is there an expectation that if user did not write to some parts of a buffer, then it must have zeros in them ? 

It's just there is a cost to clearing, and I'm trying to understand if it is needed at all ?

Reply:
Yes, this is required to keep this consistent with the temp file flow where zero bytes are present if a user doesn't write the data.

Reply:
I understand. But that just makes me question the clearing in the temp-file flow as well..

For now, could you add a comment/question here to revisit this choice ?

Reply:
Added a todo to revisit this later.